crates/snp_validation/src/lib.rs

```diff
@@ -287,96 +308,132 @@ impl Snp {
+            "snp report has an unknown version"
+        );
 
         // Check policy
-        if !report.body.policy.flags.contains(PolicyFlags::Reserved) {
-            return Err(anyhow!("snp guest policy mandatory reserved flag not set"));
-        }
-
-        if report.body.policy.flags.contains(PolicyFlags::MigrateMA) {
-            return Err(anyhow!("snp guest policy migration flag was set"));
+        ensure!(
+            report.body.policy.flags.contains(PolicyFlags::Reserved),
+            "snp guest policy mandatory reserved flag not set"
+        );
+
+        // Enarx doesn't support migration to another machine
+        ensure!(
+            !report.body.policy.flags.contains(PolicyFlags::MigrateMA),
+            "snp guest policy migration flag was set"
+        );
+
+        // Only check major.minor, as there isn't a revision or patch version in the attestation.
+        if report.body.policy.abi_major > report.body.current_major {
+            bail!("snp policy has higher abi major than firmware");
+        } else if report.body.policy.abi_minor > report.body.current_minor {
+            bail!("snp policy has higher abi minor than firmware");
         }
 
         // Check reserved fields
         if report.body.rsvd1 != 0 || report.body.rsvd3 != 0 || report.body.rsvd4 != 0 {
-            return Err(anyhow!("snp report reserved fields were set"));
+            bail!("snp report reserved fields were set");
         }
 
         for value in report.body.rsvd2 {
-            if value != 0 {
-                return Err(anyhow!("snp report reserved fields were set"));
-            }
+            ensure!(value == 0, "snp report reserved fields were set");
         }
 
         for value in report.body.rsvd5 {
-            if value != 0 {
-                return Err(anyhow!("snp report reserved fields were set"));
-            }
+            ensure!(value == 0, "snp report reserved fields were set");
         }
 
         for value in report.body.policy.rsvd {
-            if value != 0 {
-                return Err(anyhow!("snp report policy reserved fields were set"));
-            }
+            ensure!(value == 0, "snp report policy reserved fields were set");
         }
 
         for value in report.body.plat_info.rsvd {
-            if value != 0 {
-                return Err(anyhow!("snp report platform_info reserved fields were set"));
-            }
+            ensure!(
+                value == 0,
+                "snp report platform_info reserved fields were set"
+            );
         }
 
         // Check fields not set by Enarx
-        for value in report.body.author_key_digest {
-            if value != 0 {
-                return Err(anyhow!(
-                    "snp report author_key_digest field not set by Enarx"
-                ));
-            }
+        for value in report.body.host_data {
+            ensure!(value == 0, "snp report host_data field not set by Enarx");
         }
```

Comment:
This is a fixed-size byte array, just use `is_empty()` directly, there's no need for iteration. Same for most other occurrences.
See diff I posted

Reply:
`.is_empty()` would return true for an array of zeros? Doesn't seem right.

Reply:
Yes, it would, why would that not be right?

Reply:
The reason why it would not work is that arrays coerce to slices and https://doc.rust-lang.org/std/primitive.slice.html#method.is_empty checks for length of 0.
Interestingly, the test cases passed with my diff using `is_empty`, so looks like test suite is not complete.
Have you tried to check the coverage?

In any case, you should not be iterating a byte array.

Reply:
Can't use `.is_empty()`, it simply returns `self.len() == 0`, which isn't valid. These reserved fields need to be set to zeros, which means they're not empty.